File: packages/lib-shared-js/src/logging/utils/Logger/_Logger.ts
import { DateTime } from '@lib/shared/datetime/utils/DateTime/DateTime';
import { DATETIME_FORMAT } from '@lib/shared/datetime/utils/DateTime/DateTime.constants';
import {
  type _LoggerModel,
  type _LoggerParamsModel,
} from '@lib/shared/logging/utils/Logger/_Logger.models';
import { type LogArgsModel } from '@lib/shared/logging/utils/Logger/Logger.models';
import log, { type LogLevelDesc } from 'loglevel';
import prefix from 'loglevel-plugin-prefix';

export class _Logger implements _LoggerModel {
  constructor(params: _LoggerParamsModel) {
    prefix.reg(log);
    prefix.apply(log, {
      timestampFormatter: (timestamp: Date) =>
        new DateTime(timestamp).format(DATETIME_FORMAT.DATE_TIME_MS),
    });
    log.setLevel(params.level as LogLevelDesc);
  }

  debug(params: LogArgsModel, ...rest: Array<LogArgsModel>): void {
    log.debug(params, ...rest);
  }

  error(params: LogArgsModel, ...rest: Array<LogArgsModel>): void {
    log.error(params, ...rest);
  }

  info(params: LogArgsModel, ...rest: Array<LogArgsModel>): void {
    log.info(params, ...rest);
  }

  trace(params: LogArgsModel, ...rest: Array<LogArgsModel>): void {
    log.trace(params, ...rest);
  }

  warn(params: LogArgsModel, ...rest: Array<LogArgsModel>): void {
    log.warn(params, ...rest);
  }
}

File: packages/lib-shared-py/src/lib_shared/core/utils/logger/_logger.py
from logging import DEBUG, ERROR, INFO, NOTSET, WARN, Logger, getLogger, basicConfig
from typing import Any

from lib_shared.core.utils.logger._logger_models import _LoggerModel
from lib_shared.core.utils.logger.logger_constants import LoggerLevel


class _Logger(_LoggerModel):
    logger: Logger

    def __init__(self, level: LoggerLevel) -> None:
        logger_level = NOTSET
        match level:
            case LoggerLevel.DEBUG:
                logger_level = DEBUG
            case LoggerLevel.ERROR:
                logger_level = ERROR
            case LoggerLevel.INFO:
                logger_level = INFO
            case LoggerLevel.WARN:
                logger_level = WARN

        self.logger = getLogger()
        self.logger.setLevel(logger_level)
        basicConfig(
            level=logger_level,
            # format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
        )

    def debug(self, message: Any) -> None:
        self.logger.debug(message)

    def error(self, message: Any) -> None:
        self.logger.error(message)

    def info(self, message: Any) -> None:
        self.logger.info(message)
